File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/redux/ReduxFacade.java
package com.github.jmlparser.redux;

import com.github.javaparser.ast.Node;

import java.util.ArrayList;
import java.util.List;
import java.util.ServiceLoader;

/**
 * @author Alexander Weigl
 * @version 1 (12/29/21)
 */
public class ReduxFacade  {
    private final List<Transformer> transformers = new ArrayList<>();
    private final ReduxConfig config;

    public ReduxFacade(ReduxConfig config) {
        this.config = config;
    }

    public ReduxFacade(ReduxConfig config, List<Transformer> transformers) {
        this(config);
        this.transformers.addAll(transformers);
    }

    public static ReduxFacade create(ReduxConfig config) {
        var sl = ServiceLoader.load(Transformer.class);
        return new ReduxFacade(config, sl.stream()
                .filter(it -> config.isEnabled(it.type().toString()))
                .map(ServiceLoader.Provider::get).toList());
    }

    public List<? extends Node> apply(List<? extends Node> nodes) {
        List<Node> n = new ArrayList<>(nodes.size());
        for (Node node : nodes) {
            for (Transformer transformer : transformers) {
                node = transformer.apply(node);
            }
            n.add(node);
        }
        return n;
    }

}

File: tools/redux/src/main/kotlin/io/github/jmltoolkit/redux/ReduxFacade.kt
package io.github.jmltoolkit.redux


import com.github.javaparser.ast.Node
import java.util.*

/**
 * @author Alexander Weigl
 * @version 1 (12/29/21)
 */
class ReduxFacade(private val config: ReduxConfig) {
    private val transformers: MutableList<Transformer> = ArrayList()

    constructor(config: ReduxConfig, transformers: List<Transformer>) : this(config) {
        this.transformers.addAll(transformers)
    }

    fun apply(nodes: List<Node>): List<Node?> {
        val n: MutableList<Node?> = ArrayList(nodes.size)
        for (node in nodes) {
            var new = node
            for (transformer in transformers) {
                new = transformer.apply(new)
            }
            n.add(new)
        }
        return n
    }

    companion object {
        fun create(config: ReduxConfig): ReduxFacade {
            val sl = ServiceLoader.load(Transformer::class.java)
            return ReduxFacade(config, sl.filter { config.isEnabled(it.javaClass.name) })
        }
    }
}
